exchanges/order/timeinforce.go

```diff
@@ -0,0 +1,125 @@
+package order
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+)
+
+// var error definitions
+var (
+	ErrInvalidTimeInForce     = errors.New("invalid time in force value provided")
+	ErrUnsupportedTimeInForce = errors.New("unsupported time in force value")
+)
+
+// TimeInForce enforces a standard for time-in-force values across the code base.
+type TimeInForce uint16
+
+// TimeInForce types
+const (
+	UnknownTIF     TimeInForce = 0
+	GoodTillCancel TimeInForce = 1 << iota
+	GoodTillDay
+	GoodTillTime
+	GoodTillCrossing
+	FillOrKill
+	ImmediateOrCancel
+	PostOnly
+
+	supportedTimeInForceFlag = GoodTillCancel | GoodTillDay | GoodTillTime | GoodTillCrossing | FillOrKill | ImmediateOrCancel | PostOnly
+)
+
+// Is checks to see if the enum contains the flag
+func (t TimeInForce) Is(in TimeInForce) bool {
+	return in != 0 && t&in == in
+}
+
+// StringToTimeInForce converts time in force string value to TimeInForce instance.
+func StringToTimeInForce(timeInForce string) (TimeInForce, error) {
+	var result TimeInForce
+	timeInForce = strings.ToUpper(timeInForce)
+	switch timeInForce {
+	case "IMMEDIATEORCANCEL", "IMMEDIATE_OR_CANCEL", ImmediateOrCancel.String():
+		result = ImmediateOrCancel
+	case "GOODTILLCANCEL", "GOODTILCANCEL", "GOOD_TIL_CANCELLED", "GOOD_TILL_CANCELLED", "GOOD_TILL_CANCELED", GoodTillCancel.String():
+		result = GoodTillCancel
+	case "GOODTILLDAY", GoodTillDay.String(), "GOOD_TIL_DAY", "GOOD_TILL_DAY":
+		result = GoodTillDay
+	case "GOODTILLTIME", "GOOD_TIL_TIME", GoodTillTime.String():
+		result = GoodTillTime
+	case "GOODTILLCROSSING", "GOOD_TIL_CROSSING", "GOOD TIL CROSSING", GoodTillCrossing.String(), "GOOD_TILL_CROSSING":
+		result = GoodTillCrossing
+	case "FILLORKILL", "FILL_OR_KILL", FillOrKill.String():
+		result = FillOrKill
+	case PostOnly.String(), "POC", "POST_ONLY", "PENDINGORCANCEL":
+		result = PostOnly
+	case "POST_ONLY_GOOD_TIL_CANCELLED":
+		result = GoodTillCancel | PostOnly
+	}
+	if result == UnknownTIF && timeInForce != "" {
+		return UnknownTIF, fmt.Errorf("%w: tif=%s", ErrInvalidTimeInForce, timeInForce)
+	}
+	return result, nil
+}
+
+// IsValid returns whether or not the supplied time in force value is valid or
+// not
+func (t TimeInForce) IsValid() bool {
+	// Neither ImmediateOrCancel nor FillOrKill can coexist with anything else
+	// If either bit is set then it must be the only bit set
+	if t&(ImmediateOrCancel|FillOrKill) != 0 && t&(t-1) != 0 {
+		return false
+	}
+	return t == UnknownTIF || supportedTimeInForceFlag&t == t
```

Comment:
🐒 Nitpick
Sorry; I looked at this again and found that everyone is still going to try to parse the code, and it's still an eyesore.
I think this is optimal for readability / maintenance:
```suggestion
	// Neither ImmediateOrCancel nor FillOrKill can coexist with anything else
	// If either bit is set then it must be the only bit set
	isIOCorFOK := t&(ImmediateOrCancel|FillOrKill) != 0
	hasTwoBitsSet := t&(t-1) != 0
	if isIOCorFOK && hasTwoBitsSet {
		return false
	}
```

🥃 Whilst this is cool, I don't think it's legible enough:
```
	// Neither ImmediateOrCancel nor FillOrKill can coexist with anything else
	// If either bit is set then it must be the only bit set
	if isIOCorFOK, hasTwoBitsSet := t&(ImmediateOrCancel|FillOrKill) != 0, t&(t-1) != 0; isIOCorFOK && hasTwoBitsSet {
		return false
	}
```